src/combinatorics/simple_graph/basic.lean

```diff
@@ -122,13 +149,63 @@ set.to_finset G.edge_set
   e ∈ G.edge_finset ↔ e ∈ G.edge_set :=
 by { dunfold edge_finset, simp }
 
+@[simp] lemma edge_set_univ_card [decidable_eq V] [fintype V] [decidable_rel G.adj] :
+  (univ : finset G.edge_set).card = G.edge_finset.card :=
+fintype.card_of_subtype G.edge_finset (mem_edge_finset _)
+
 @[simp] lemma irrefl {v : V} : ¬G.adj v v := G.loopless v
 
 @[symm] lemma edge_symm (u v : V) : G.adj u v ↔ G.adj v u := ⟨λ x, G.sym x, λ x, G.sym x⟩
 
 @[simp] lemma mem_neighbor_set (v w : V) : w ∈ G.neighbor_set v ↔ G.adj v w :=
 by tauto
 
+@[simp] lemma mem_incidence_set (v w : V) : ⟦(v, w)⟧ ∈ G.incidence_set v ↔ G.adj v w :=
+by { dsimp [incidence_set], simp }
+
+lemma neighbor_set_edge_prop {v w : V} (h : w ∈ G.neighbor_set v) : ⟦(v, w)⟧ ∈ G.incidence_set v :=
+by { rw mem_neighbor_set at h, simpa }
+
+lemma adj_incidence_set_inter {v : V} {e : sym2 V} (he : e ∈ G.edge_set) (h : v ∈ e) :
+  G.incidence_set v ∩ G.incidence_set h.other = {e} :=
+begin
+  ext e',
+  simp only [incidence_set, set.mem_sep_eq, set.mem_inter_eq, set.mem_singleton_iff],
+  split,
+  { intro h', rw ←sym2.mem_other_spec h,
+    exact (sym2.elems_iff_eq (edge_other_ne G he h).symm).mp ⟨h'.1.2, h'.2.2⟩, },
+  { rintro rfl, use [he, h, he], apply sym2.mem_other_mem, },
+end
+
+section incidence
+variable [decidable_eq V]
+
+/--
+Given an edge incident to a particular vertex, get the other vertex on the edge.
+-/
+def incidence_set_other {v : V} {e : sym2 V} (h : e ∈ G.incidence_set v) : V := h.2.other'
+
+lemma incidence_other_prop {v : V} {e : sym2 V} (h : e ∈ G.incidence_set v) :
+G.incidence_set_other h ∈ G.neighbor_set v :=
+by { cases h, rwa [←sym2.mem_other_spec' h_right, mem_edge_set] at h_left }
+
+@[simp]
+lemma incidence_other_neighbor_edge {v w : V} (h : w ∈ G.neighbor_set v) :
+  G.incidence_set_other (G.neighbor_set_edge_prop h) = w :=
+sym2.congr_right.mp (sym2.mem_other_spec' (G.neighbor_set_edge_prop h).right)
+
+/--
+There is an equivalence between the set of edges incident to a given
+vertex and the set of vertices adjacent to the vertex.
+-/
+@[simps] def incidence_set_equiv_neighbor_set (v : V) : G.incidence_set v ≃ G.neighbor_set v :=
+{ to_fun := λ e, ⟨G.incidence_set_other e.2, G.incidence_other_prop e.2⟩,
+  inv_fun := λ w, ⟨⟦(v, w.1)⟧, G.neighbor_set_edge_prop w.2⟩,
+  left_inv := by { intro x, dsimp [incidence_set_other], simp },
+  right_inv := by { intro x, rcases x with ⟨w, hw⟩, simp, } }
```

Comment:
With the above suggestion to make it an iff, this becomes (with a little golfing)
```lean
@[simps] def incidence_set_equiv_neighbor_set (v : V) : G.incidence_set v ≃ G.neighbor_set v :=
{ to_fun := λ e, ⟨G.incidence_set_other e.2, G.incidence_other_prop e.2⟩,
  inv_fun := λ w, ⟨⟦(v, w.1)⟧, G.mem_incident_iff_neighbor.mpr w.2⟩,
  left_inv := λ x, by simp [incidence_set_other],
  right_inv := λ ⟨w, hw⟩, by simp }
```